### PR TITLE
feat: drop Has* interface usage to ease fx adoption

### PR DESCRIPTION
**Pull request by sethyates**
## Description

Drop Has* interface usage to ease fx adoption

## Motivation and Context

Slowly migrating to cleaner fx integration

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by sethyates** [comment]
Also drops the deprecated loopback GRPC-JSON gateway

**Review by algorithmcardboard — COMMENTED** [review]
Shouldn't this commit be prefixed with "Breaking change" as well?

File: http.go [hunk 1]
```diff
@@ -50,15 +50,15 @@ func NewHttpClient(ctx context.Context, cfg HasClientConfig, vault HasVaultConfi
...
+				IdleConnTimeout: cfg.ConnTimeout,
+				WriteBufferSize: cfg.WriteBufferSize,
+				ReadBufferSize:  cfg.ReadBufferSize,
 			},
 		},
 	}
 
-	if cfg.GetToken() != nil && cfg.GetToken().GetShared() != nil {
-		if httpCli.token, err = LoadKey(ctx, cfg.GetToken().GetShared(), vault, "shared"); err != nil {
+	if cfg.Token.Shared.GetEnabled() {
```
Comment (algorithmcardboard) [inline http.go:60]: Dont we need a null check for cfg.Token?

File: credentials.go [hunk 2]
```diff
@@ -27,62 +27,40 @@ import (
...
 		return &Credentials{
-			Username: cfg.GetUsername(),
-			Password: cfg.GetPassword(),
+			Username: cfg.Username,
+			Password: cfg.Password,
 		}, nil
 	}
 
-	if len(cfg.GetID()) == 0 || vault == nil || !vault.GetEnabled() {
+	if len(cfg.ID) == 0 || !vault.GetEnabled() {
```
Comment (algorithmcardboard) [inline credentials.go:63]: Will this panic if vault is null?  We do have a null check previously?

File: key.go [hunk 3]
```diff
@@ -34,37 +34,37 @@ import (
...
-		fields["key.file"] = cfg.GetFile()
-		span.SetAttributes(attribute.String("key.file", cfg.GetFile()))
+		fields["key.file"] = cfg.File
+		span.SetAttributes(attribute.String("key.file", cfg.File))
 	}
 
-	if len(cfg.GetID()) > 0 {
-		if vault != nil && vault.GetEnabled() {
+	if len(cfg.ID) > 0 {
+		if vault.GetEnabled() {
```
Comment (algorithmcardboard) [inline key.go:60]: can vault be null here?

**Comment by sethyates** [comment]
Theoretically, but that would bump to v2 and that requires a huge amount of changes. I've opened other PR's that fix any breaks.

**Review by algorithmcardboard — APPROVED** [review]
The changes look good except for the null checks.  Since I dont have much context around the null checks wrt to the flows we are calling them, I am okay for these changes to be merged if you think the null checks wouldn't be a problem.

File: key.go (hunk 3 above)
Comment (sethyates) [reply key.go:60]: Not anymore since we pass config by value instead of using the interface.

File: credentials.go (hunk 2 above)
Comment (sethyates) [reply credentials.go:63]: No longer possible to be nil.

File: http.go (hunk 1 above)
Comment (sethyates) [reply http.go:60]: No longer possible to be nil.

**Pull request merged**